source/pervasive/atomic.rs

```diff
@@ -0,0 +1,468 @@
+            opens_invariants_none();
+
+            return self.ato.fetch_and(n, Ordering::SeqCst);
+        }
+
+        #[inline(always)]
+        #[verifier(external_body)]
+        #[verifier(atomic)]
+        pub fn fetch_or(&self, #[proof] perm: &mut $p_ident, n: $value_ty) -> $value_ty {
+            requires([
+                equal(self.view(), old(perm).patomic),
+            ]);
+            ensures(|ret: $value_ty| equal(old(perm).value, ret)
+                && perm.patomic == old(perm).patomic
+                && perm.value == (old(perm).value || n)
+            );
+            opens_invariants_none();
+
+            return self.ato.fetch_or(n, Ordering::SeqCst);
+        }
+
+        #[inline(always)]
+        #[verifier(external_body)]
+        #[verifier(atomic)]
+        pub fn fetch_xor(&self, #[proof] perm: &mut $p_ident, n: $value_ty) -> $value_ty {
+            requires([
+                equal(self.view(), old(perm).patomic),
+            ]);
+            ensures(|ret: $value_ty| equal(old(perm).value, ret)
+                && perm.patomic == old(perm).patomic
+                && perm.value == ((old(perm).value && !n) || (!old(perm).value && n))
+            );
+            opens_invariants_none();
+
+            return self.ato.fetch_or(n, Ordering::SeqCst);
+        }
+
+        #[inline(always)]
+        #[verifier(external_body)]
+        #[verifier(atomic)]
+        pub fn fetch_nand(&self, #[proof] perm: &mut $p_ident, n: $value_ty) -> $value_ty {
+            requires([
+                equal(self.view(), old(perm).patomic),
+            ]);
+            ensures(|ret: $value_ty| equal(old(perm).value, ret)
+                && perm.patomic == old(perm).patomic
+                && perm.value == !(old(perm).value && n)
+            );
+            opens_invariants_none();
+
+            return self.ato.fetch_nand(n, Ordering::SeqCst);
+        }
+    }
+}
+
+make_bool_atomic!(PAtomicBool, PermissionBool, AtomicBool, bool);
+
+make_integer_atomic!(PAtomicU8, PermissionU8, AtomicU8, u8, wrapping_add_u8, wrapping_sub_u8, 0, 0xff);
+make_integer_atomic!(PAtomicU16, PermissionU16, AtomicU16, u16, wrapping_add_u16, wrapping_sub_u16, 0, 0xffff);
+make_integer_atomic!(PAtomicU32, PermissionU32, AtomicU32, u32, wrapping_add_u32, wrapping_sub_u32, 0, 0xffff_ffff);
+make_integer_atomic!(PAtomicU64, PermissionU64, AtomicU64, u64, wrapping_add_u64, wrapping_sub_u64, 0, 0xffff_ffff_ffff_ffff);
+
+make_integer_atomic!(PAtomicI8, PermissionI8, AtomicI8, i8, wrapping_add_i8, wrapping_sub_i8, -0x80, 0x7f);
+make_integer_atomic!(PAtomicI16, PermissionI16, AtomicI16, i16, wrapping_add_i16, wrapping_sub_i16, -0x8000, 0x7fff);
+make_integer_atomic!(PAtomicI32, PermissionI32, AtomicI32, i32, wrapping_add_i32, wrapping_sub_i32, -0x8000_0000, 0x7fff_ffff);
+make_integer_atomic!(PAtomicI64, PermissionI64, AtomicI64, i64, wrapping_add_i64, wrapping_sub_i64, -0x8000_0000_0000_0000, 0x7fff_ffff_ffff_ffff);
+
+// TODO usize and isize
```

Comment:
Are you missing something for these, or just planning to implement them later?

Reply:
I can't implement them until we have a way to use constants like `usize::MAX`. I'll add a note